wagtail/wagtailcore/models.py

```diff
@@ -40,8 +39,86 @@
 
 PAGE_TEMPLATE_VAR = 'page'
 
+SITE_CACHE = {}
+REQUEST_SITE_MATCH_CACHE = {}
+
 
 class SiteManager(models.Manager):
+
+    @staticmethod
+    def get_hostname_and_port_from_request(request):
+        """Reliably extract 'hostname' and 'port' values from the supplied
+        ``HttpRequest`` instance (without raising exceptions), and return them
+        as a tuple.
+        """
+        try:
+            hostname = request.get_host().split(':')[0]
+        except KeyError:
+            hostname = None
+        try:
+            port = request.get_port()
+        except KeyError:
+            port = None
+        return hostname, port
+
+    @staticmethod
+    def clear_site_cache():
+        global SITE_CACHE
+        global REQUEST_SITE_MATCH_CACHE
+        SITE_CACHE = {}
+        REQUEST_SITE_MATCH_CACHE = {}
+
+    def _get_by_id(self, site_id):
+        if site_id not in SITE_CACHE:
+            site = self.get(pk=site_id)
+            SITE_CACHE[site_id] = site
+        return SITE_CACHE[site_id]
+
+    def _cache_and_return_match(self, site, match_cache_key):
+        SITE_CACHE[site.pk] = site
+        REQUEST_SITE_MATCH_CACHE[match_cache_key] = site.pk
+        return SITE_CACHE[site.pk]
+
+    def get_default(self):
+        """Returns the 'default' ``Site`` or raise an exception in no site is
+        set as the default. The result is cached."""
+        key = 'default'
+        if key in REQUEST_SITE_MATCH_CACHE:
+            return self._get_by_id(REQUEST_SITE_MATCH_CACHE['default'])
+
+        match = self.get(is_default_site=True)
+        return self._cache_and_return_match(match, key)
```

Comment:
When caching the default site, you could save it both in the `'__default__'` key and the one of its host+port. This will avoid extra SQL queries when fetching that site again for the first time.

Reply:
Great suggestion.

Reply:
Oh, wait. This is already what happens in the last line of `get_for_request()`